### PR TITLE
Make functorch compilable with Py-3.11

### LINKED ISSUE #85006: Python 3.11 Wheels are broken by functorch commit
### 🐛 Describe the bug

We have all linux binary 3.11 wheel builds broken in nightly workflow:
https://hud.pytorch.org/hud/pytorch/pytorch/nightly

Example breakage:
https://github.com/pytorch/pytorch/actions/runs/3050769707/jobs/4918244719

Suspect is this PR:
https://github.com/pytorch/pytorch/pull/83464

### Versions

nightly 1.13

cc @seemethere @malfet @pytorch/pytorch-dev-infra @zou3519 @Chillee @samdow @soumith

### PR DESCRIPTION
**Pull request by malfet**
By using compatibility wrappers from [python_compat.h](https://github.com/pytorch/pytorch/blob/master/torch/csrc/utils/python_compat.h) and skipping part of `getname` switch

Fixes https://github.com/pytorch/pytorch/issues/85006

Please note that `import torch` right now fails by default on 3.11 with some jit issue, so I think this shouldn't be a really issue for a bit

### CODE REVIEW

**Review by kit1980 — APPROVED** [review]

**Comment by zou3519** [comment]
Test failures look legit

**Comment by malfet** [comment]
> Test failures look legit
Yeah, because wrappers/3.10+ code takes strong references, but current dims.cpp assumes they are weak ones. Modified

File: functorch/csrc/dim/dim.cpp [hunk 1]
```diff
@@ -1500,13 +1511,15 @@ static PyObject* _dims(PyObject *self,
     }
 
     PyThreadState* state = PyThreadState_GET();
-    PyFrameObject* f = state->frame;
-    auto code = (_Py_CODEUNIT*)PyBytes_AS_STRING(f->f_code->co_code);
+    PyFrameObject* f = PyThreadState_GetFrame(state);
```
Comment (zdevito) [inline functorch/csrc/dim/dim.cpp:1514]: These can use `py::object::steal` to keep ownership in an object rather than the explicit Py_DECREF:
```
auto frame_owner = py::object::steal((PyObject*) f);
```

File: functorch/csrc/dim/dim.cpp [hunk 2]
```diff
@@ -1439,12 +1440,15 @@ py::object getname(PyCodeObject* code, _Py_CODEUNIT c) {
         case STORE_GLOBAL:
           names = code->co_names;
           break;
+#if PY_VERSION_HEX < 0x030b0000
+        // TODO: Re-enable this for Python-3.11
```
Comment (zdevito) [inline functorch/csrc/dim/dim.cpp:1444]: `PyCode_GetVarnames` and `PyCode_GetCellVars` were added. By changing names to a `py::object` to retain ownership, this fix is a couple lines.

**Review by zdevito — APPROVED** [review]
Looks good, thanks for fixing this!

File: functorch/csrc/dim/dim.cpp (hunk 2 above)
Comment (malfet) [reply functorch/csrc/dim/dim.cpp:1444]: Thank you for the feedback. I think a bit more fundamental problem, is that code[GetLasti(frame]) returns some strange opcodes (although dis shows that it still the same)

Comment (malfet) [reply functorch/csrc/dim/dim.cpp:1444]: Actually, this might be because we are testing everything using beta1 build of 3.11, let's see if updating to rc2 will make a difference: https://github.com/pytorch/builder/pull/1149
